File: 2095. Delete the Middle Node of a Linked List.py
class Solution:
    def deleteMiddle(
        self,
        head_node: Optional[ListNode]
    ) -> Optional[ListNode]:

        # If there is only one node,
        # deleting the middle makes the list empty
        if not head_node or not head_node.next:
            return None

        slow_pointer = head_node
        fast_pointer = head_node
        previous_node = None

        # Find the middle node
        while fast_pointer and fast_pointer.next:

            previous_node = slow_pointer

            slow_pointer = slow_pointer.next

            fast_pointer = (
                fast_pointer.next.next
            )

        # Delete the middle node
        previous_node.next = (
            slow_pointer.next
        )

        return head_node

File: 2095. Delete the Middle Node of a Linked List.java
class Solution {

    public ListNode deleteMiddle(
            ListNode headNode) {

        // If there is only one node
        if (headNode == null ||
            headNode.next == null) {

            return null;
        }

        ListNode slowPointer =
                headNode;

        ListNode fastPointer =
                headNode;

        ListNode previousNode =
                null;

        // Find the middle node
        while (fastPointer != null &&
               fastPointer.next != null) {

            previousNode =
                    slowPointer;

            slowPointer =
                    slowPointer.next;

            fastPointer =
                    fastPointer.next.next;
        }

        // Delete the middle node
        previousNode.next =
                slowPointer.next;

        return headNode;
    }
}
